internal/cmd/sshauthkeys.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/infrahq/infra/internal/linux"
+)
+
+func newSSHDCmd(cli *CLI) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:    "sshd",
+		Short:  "Commands for integrating with ssh server",
+		Hidden: true,
+	}
+
+	cmd.AddCommand(newSSHDAuthKeysCmd(cli))
+	return cmd
+}
+
+type sshAuthKeysOptions struct {
+	fingerprint    string
+	username       string
+	configFilename string
+}
+
+func newSSHDAuthKeysCmd(cli *CLI) *cobra.Command {
+	var opts sshAuthKeysOptions
+	cmd := &cobra.Command{
+		Use:    "auth-keys",
+		Hidden: true,
+		Args:   ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logger := setupLogger(cli)
+
+			// sshd_config: AuthorizedKeysCommand infra ssh auth-keys %u %f
+			opts.username = args[0]
+			opts.fingerprint = args[1]
+
+			// log the error to syslog, since we expect stdout/stderr to be hidden
+			if err := runSSHDAuthKeys(cli, logger, opts); err != nil {
+				logger.Err(err).Msg("ssh auth-keys exit")
+				return err
+			}
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.configFilename, "config-file",
+		"/etc/infra/connector.yaml", "Path to connector config file")
+	return cmd
+}
+
+func setupLogger(cli *CLI) zerolog.Logger {
+	out := []io.Writer{cli.Stderr}
+	// TODO: log to stderr if this fails?
+	syslog, _ := logsyslog.New(logsyslog.LOG_AUTH|logsyslog.LOG_WARNING, "infra-ssh")
+	if syslog != nil {
+		out = append(out, zerolog.SyslogLevelWriter(syslog))
+	}
+	return zerolog.New(zerolog.MultiLevelWriter(out...))
+}
+
+func runSSHDAuthKeys(cli *CLI, logger zerolog.Logger, opts sshAuthKeysOptions) error {
+	ctx := context.Background()
+	logger.Debug().
+		Str("username", opts.username).
+		Str("fingerprint", opts.fingerprint).
+		Msgf("Running infra ssh auth-keys")
+
+	// This command only uses a small subset of these options, but its expected
+	// that it runs in the same environment as the ssh connector, so may as well
+	// share the config.
+	config := defaultConnectorOptions()
+	err := cliopts.Load(&config, cliopts.Options{
+		Filename:  opts.configFilename,
+		EnvPrefix: "INFRA_CONNECTOR",
+	})
+	if err != nil {
+		return err
+	}
+
+	client := config.APIClient()
+	client.Name = "ssh-auth-keys-cmd"
+
+	user, err := verifyUsernameAndFingerprint(ctx, logger, client, opts)
+	if err != nil {
+		return err
+	}
+
+	if err := verifyUserIsManagedByInfra(opts.username); err != nil {
+		return err
+	}
+
+	if err := authorizeUserForDestination(ctx, client, user, config.Name); err != nil {
+		return err
+	}
+
+	for _, key := range user.PublicKeys {
+		// TODO: add expiration to this output when it's available
+		cli.Output("%v %v", key.KeyType, key.PublicKey)
+		logger.Debug().Msgf("%v %v", key.KeyType, key.PublicKey)
+	}
+	return nil
+}
+
+func verifyUsernameAndFingerprint(
+	ctx context.Context,
+	logger zerolog.Logger,
+	client *api.Client,
+	opts sshAuthKeysOptions,
+) (*api.User, error) {
+	users, err := client.ListUsers(ctx, api.ListUsersRequest{
```

Comment:
Does this mean the connector has to call the server every time it verifies a username/fingerprint? Does this happen frequently or just when a session is opened?

Reply:
Not the connector, but `infra sshd auth-keys` does yes. It will make two API calls per connect , one for the user, and one for grants. It only happens when a new SSH connection is opened.